src/validation/schema.js

```diff
@@ -322,13 +378,200 @@ const validateEntityFields = (defs, def) =>
+          message: '@imports type objects accept and require two fields: [name, as]',
+        }),
+      )
+    }
+    return type.fields.reduce((errors, field) => {
+      if (!['name', 'as'].includes(field.name.value)) {
+        return errors.push(
+          immutable.fromJS({
+            loc: directive.name.loc,
+            entity: def.name.value,
+            message: `@imports type object field '${field.name.value}' invalid,  may only be one of: [name, as]`,
+          }),
+        )
+      }
+      if (field.value.kind != 'StringValue') {
+        return errors.push(
+          immutable.fromJS({
+            loc: directive.name.loc,
+            entity: def.name.value,
+            message: '@imports type object fields [name, as] must be strings',
+          }),
+        )
+      }
+      return errors
+    }, errors)
+  },
+}
+
+const validateImportDirectiveType = (def, directive, type) =>
+  importDirectiveTypeValidators[type.kind]
+    ? importDirectiveTypeValidators[type.kind](def, directive, type)
+    : List().push(
+        immutable.fromJS({
+          loc: directive.name.loc,
+          entity: def.name.value,
+          message: 'Imported type must be either String | {name: String, as: String}',
+        }),
+      )
+
+const validateImportDirectiveArgumentTypesIsValid = (def, directive, argument) => {
+  if (argument.value.kind != 'ListValue') {
+    return List().push(
+      immutable.fromJS({
+        loc: directive.name.loc,
+        entity: def.name.value,
+        message: '@imports directive argument: types must be an list',
+      }),
+    )
+  }
+
+  return argument.value.values.reduce(
+    (errors, type) => errors.concat(validateImportDirectiveType(def, directive, type)),
+    List(),
+  )
+}
+
+const validateImportDirectiveArgumentFromIsValid = (def, directive, argument) => {
+  if (argument.value.kind != 'ObjectValue') {
+    return List().push(
+      immutable.fromJS({
+        loc: directive.name.loc,
+        entity: def.name.value,
+        message: '@imports directive argument: `from` must be an object',
+      }),
+    )
+  }
+
+  if (argument.value.fields.length != 1) {
+    return List().push(
+      immutable.fromJS({
+        loc: directive.name.loc,
+        entity: def.name.value,
+        message: `@imports directive argument: 'from' must have one field of id or name`,
+      }),
+    )
+  }
+
+  return argument.value.fields.reduce((errors, field) => {
+    if (!['name', 'id'].includes(field.name.value)) {
+      return errors.push(
+        immutable.fromJS({
+          loc: field.name.loc,
+          entity: def.name.value,
+          message: `@imports directive from: only fields 'id' or 'name' allowed`,
+        }),
+      )
+    }
+    if (field.value.kind != 'StringValue') {
+      return errors.push(
+        immutable.fromJS({
+          loc: field.name.loc,
+          entity: def.name.value,
+          message: '@imports directive from: fields must have string values',
```

Comment:
In the `types` validation you mentioned which fields there are. I think it would be good to be specific here as well if we can.

Reply:
Also: `@imports directive from:` -> `@import argument 'from': ....`?